Implement project overview page

TheRealJon opened the pull request:
Basic project overview implementation with grouping and filtering of resources. Deployment Configs, Deployments, and Stateful Sets are currently the only top-level resources shown.

![image](https://user-images.githubusercontent.com/22625502/44527998-6f844c00-a6b6-11e8-9135-66a77da47b91.png)


@openshift/team-ux-review

benjaminapetersen commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,341 @@
+const GRAPH_VIEW = 'graph';
+
+const ProjectOverviewToolbar = ({activeView, setView}) => {
+  return <Toolbar>
```
@sg00dwin ping!
I know you are consolidating a handful of toolbars throughout the console, would be good to take a quick look at this for future consistency.

benjaminapetersen commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,341 @@
+  </Toolbar>;
+};
+
+const DeploymentConfigOverview = ({obj}) => {
```
I'm inclined to split this out early into separate component files, as well as see how generic they can be (perhaps smaller pieces will allow for generic lists with some nested child components for customization, perhaps imported from our existing `deployment-config.tsx`, etc files, not sure).  

Thinking about structure early for deduplication sake.

benjaminapetersen replied:
Ah, I see you are doing some of that already, nice.

TheRealJon replied:
Yes, plan on doing similar breaking down for other resources that will be displayed in this view, but just haven't gotten there yet. I wanted to put up a PR so that others could see what I am working on and start getting feedback early, but there's still work to be done.

jeff-phillips-18 commented on frontend/package.json:
```diff
@@ -66,6 +66,7 @@
     "js-yaml": "3.x",
     "lodash-es": "4.x",
     "patternfly": "^3.45.0",
+    "patternfly-react": "^2.11.2",
```
We should go to 2.12, that will reduce the size cost of including pf-react since with that release it no longer pulls in all of react-bootstrap (only the components actually used).

benjaminapetersen replied:
+1 for not importing the whole world!

alecmerdler commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,341 @@
+import * as _ from 'lodash-es';
+import * as classNames from 'classnames';
+import { Button, EmptyState, Icon, Toolbar } from 'patternfly-react';
+import { Helmet } from 'react-helmet';
```
Newline between `node_modules` imports and project local imports for consistency.

benjaminapetersen commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,341 @@
+      deploymentConfig.prevController = _.size(ordered) < 2 ? null : ordered[1];
+
+      // Get the services for this deployment config
+      let configServices = [];
```
Nit, I'd suggest a `const configServices = _.reduce(allServices, () => {}, [])` to make these lists.

TheRealJon commented:
Added grouping functionality and adjusted styling according to style docs.

![image](https://user-images.githubusercontent.com/22625502/44489558-ca239680-a629-11e8-8e75-4bb71da93ce5.png)

TheRealJon commented:
@spadgett do we want to call this the first iteration? Or should I continue iterating on this PR to get the details panel working?

spadgett commented:
@TheRealJon I'd suggest we integrate this in phases. Let's work to get what you have merged.

spadgett commented:
jenkins rebuild

spadgett commented on frontend/public/components/deployment-config.tsx:
```diff
@@ -47,52 +47,56 @@ const menuActions = [
 ];
 
-export const DeploymentConfigsDetails: React.SFC<{obj: any}> = ({obj: deploymentConfig}) => {
+export const DeploymentConfigOverviewList = ({deploymentConfig}) => {
```
Let's pull this out of the PR for the moment and work on the details sidebar in a separate PR

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+  <Toolbar className="overview-toolbar">
+    <Toolbar.RightContent>
+      <div className="form-group overview-toolbar__form-group">
+        <label style={{marginRight: '10px'}}>
```
Avoid inline styles

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+        </label>
+        <Dropdown
+          className="overview-toolbar__dropdown"
+          disabled={!_.size(groupOptions)}
```
```
disabled={_.isEmpty(groupOptions)}
```

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+          disabled={!_.size(groupOptions)}
+          items={groupOptions}
+          onChange={handleGroupChange}
+          style={{display: 'inline-block'}}
```
Avoid inline styles

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+          onChange={handleFilterChange}
+        />
+      </div>
+      { showViewSelector &&
```
I would take this out of this PR. We can introduce it when we're ready to integrate topology view

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+      const leftVersion = parseInt(_.get(left, 'openshift.io/deployment-config.latest-version'), 10);
+      const rightVersion = parseInt(_.get(right, 'openshift.io/deployment-config.latest-version'), 10);
+      // Fall back to sorting by name if no deployment versions.
+      let leftName, rightName;
```
You can make this `const` and declare inside the `if`

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+    return _.toArray(replicationControllers).sort(compareDeployments);
+  }
+
+  filterItems(items) {
```
We should be using `fuzzy` to filter things for consistency with other views.

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+
+    const filterString = filterValue.toLowerCase();
+    return _.filter(items, item => {
+      const name = _.get(item, 'metadata.name', '');
```
Names have to be lowercase, so we can avoid `toLowerCase`

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+
+  getLabelGroupOptions(items) {
+    const {groupOptions} = this.state;
+    return _.chain(items)
```
Avoid `_.chain`... I don't think it works with tree shaking

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+          [key]: key
+        };
+      }, groupOptions)
+      .value();
```
Wondering if this is easier if we use `Set`

TheRealJon replied:
This is used in Dropdown. If dropdown accepts a Set or we refactor it to accept a Set, that would work.

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+      const currentController = _.head(orderedReplicationControllers) || _.head(ownedReplicaSets);
+      const prevController = orderedReplicationControllers[1];
+
+      // Get the routes for this deployment config
```
I would take this out for now if we're not showing routes or services yet. (Save the code in a branch :)

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+      // Get the routes for this deployment config
+      const templateLabel = _.get(rootResource, `spec.template.metadata.labels.${kind.toLowerCase()}`);
+      const associatedRoutes = _.reduce(services.data, (accumulator, service) => {
+        const serviceSelector = _.get(service, `spec.selector.${kind.toLowerCase()}`);
```
I'm not sure the label logic is correct. Label matching is pretty complicated.

What is `kind` here?

TheRealJon replied:
`kind` would be one of 'DeploymentConfig', 'Deployment', or 'StatefulSet' here.

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,404 @@
+    const {deploymentConfigs, deployments, loaded, statefulSets} = this.props;
+    const {selectedGroupLabel} = this.state;
+
+    if (!loaded) {
```
We might look at enabling progressive loading, but fine for now.

spadgett commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,74 @@
+        />
+      </div>
+    }
+    { latestVersion &&
```
We could consider leaving this out since it's part of the rc/rs name

TheRealJon replied:
Yeah, I thought about that too. My only (very minor) concern is that we are leaving it up to the user to know that the version number is in the rc/rs name. But I don't think that's a huge deal because it's pretty obvious.

spadgett commented on frontend/public/components/utils/firehose.jsx:
```diff
@@ -174,7 +174,7 @@ export const Firehose = connect(
           return true;
         }
-        return false;
+        return nextProps.forceUpdate;
```
Why is this needed?

TheRealJon replied:
Basically, Firehose will only re-render when the underlying resources change. So switching the view component rendered inside the Firehose will not cause a re-render. Since we are taking out the view switching functionality, this probably isn't necessary right now. We could also probably figure out a better way to do this. I just hacked something together to work temporarily.

spadgett commented on frontend/public/components/utils/resource-icon.tsx:
```diff
@@ -8,8 +8,7 @@ import { kindToAbbr } from '../../module/k8s/get-resources';
 
-export const ResourceIcon = (props: ResourceIconProps) => {
-  const { kind, className } = props;
+export const ResourceIcon: React.SFC<ResourceIconProps>= ({className, kind}) => {
```
style nit: space before `=`

spadgett commented on frontend/yarn.lock:
```diff
@@ -33,17 +33,203 @@
   version "0.7.0"
   resolved "https://registry.yarnpkg.com/@sindresorhus/is/-/is-0.7.0.tgz#9a06f4f137ee84d7df0460c1fdb1135ffa6c50fd"
 
+"@types/c3@^0.6.0":
```
A ton of dependencies are getting added here... Per @jeff-phillips-18, patternfly-react should be doing tree shaking. We should verify we're only loading what we're using.

spadgett replied:
Best I can tell, all of these dependencies are getting included in the vendor bundle (c3.js, etc.) for overview. It's almost doubling the size from our main vendor bundle. We're only using a couple components. (This is running in development mode. I'll also check the built image.)

@jeff-phillips-18 any ideas?

![project overview okd 2018-08-24 09-45-44](https://user-images.githubusercontent.com/1167259/44587989-898b6080-a782-11e8-8d5e-36daf7073ee6.png)

<img width="903" alt="project overview okd 2018-08-24 09-42-50" src="https://user-images.githubusercontent.com/1167259/44587939-5fd23980-a782-11e8-9928-78c7628d9493.png">

jeff-phillips-18 replied:
Development mode does not tree shake

spadgett replied:
I was seeing it in the built image, too, but I think I didn't correctly build it. Thanks @jeff-phillips-18, I think we're OK then.

spadgett replied:
OK guys. Now that I built it correctly, looks like we're good. Sorry for the false alarm!

![project overview 2018-08-24 10-13-15](https://user-images.githubusercontent.com/1167259/44589367-69f63700-a786-11e8-85cc-3ed3462ed7f5.png)

benjaminapetersen replied:
Curious about that. Is there any concern we will forget about the tree shaking in dev mode & accidentally ship a large bundle to prod in the future?  Or is there any risk that dev mode might load something that won't exist later, meaning we won't see the error now?  Just curious about the possibilities here.

spadgett replied:
The tree shaking stuff should just work... And with code splitting across routes, it should protect us as well.

alecmerdler replied:
The "Performance" e2e tests run against the production build, so we are covered from regressions there.

spadgett commented:
I'm happy we're able to get patternfly-react integrated

TheRealJon commented:
jenkins rebuild

spadgett commented:
jenkins push

sg00dwin commented:
@TheRealJon Here's a screenshot of the empty dropdown filter state that I mentioned.
<img width="1067" alt="project-overview" src="https://user-images.githubusercontent.com/1874151/44862049-e8038380-ac47-11e8-9019-4ef867e3678f.png">

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_393_build_417`

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,353 @@
+    }
+  }
+
+  sortByDeploymentVersion(replicationControllers, descending) {
```
It's not clear to me why we need to sort by deployment version if we're only showing the active one. We might be able to skip this for now?

TheRealJon replied:
My thought was that we will have to use some logic similar to this in order to find the replication controller with the latest version. Since we will eventually be showing the whole list, sorting and grabbing the head seems like a good way to do it to make things easier for showing the whole list. We could make this into something like 'getLatestDeploymentVersion' but we would still have to iterate over all the RC's and do this same logic to return one RC.

spadgett replied:
You should be able to look at the property on the deployment config or deployment to get the current revision. rc-10 might be failed and rc-9 is what is running... It's not necessarily the most recent by number

TheRealJon replied:
Gotcha, okay, that makes sense

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,353 @@
+import { ProjectOverview } from './project-overview';
+
+const getOwnedResources = (resources, uid) => {
+  return _.filter(resources, function (resource) {
```
nit: prefer fat arrow

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,353 @@
+  sortByDeploymentVersion(replicationControllers, descending) {
+    const compareDeployments = (left, right) => {
+      const leftVersion = parseInt(_.get(left, 'openshift.io/deployment-config.latest-version'), 10);
+      const rightVersion = parseInt(_.get(right, 'openshift.io/deployment-config.latest-version'), 10);
```
Doesn't this need to be checking the annotations? Looks like the `_.get` call is wrong.

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,353 @@
+
+    const deploymentConfigItems = this.buildGraphForRootResources(deploymentConfigs.data, 'DeploymentConfig');
+    const deploymentItems = this.buildGraphForRootResources(deployments.data, 'Deployment');
+    const statefulSetItems = this.buildGraphForReplicators(statefulSets.data, 'StatefulSet', true);
```
Looks like the method only takes two args?

TheRealJon replied:
Oops, an artifact left from a previous implementation

spadgett commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,103 @@
+const ProjectOverviewListItem = ({obj}) => {
+  const {currentController, kind, metadata} = obj;
+  const {namespace, name, uid} = metadata;
+  const kindObj = modelFor(kind);
```
We should ideally be using `connectToModels` here

spadgett commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,103 @@
+  const additionalInfo = <div key={uid} className="project-overview__additional-info">
+    { currentController &&
+      <div className="project-overview__detail project-overview__detail--controller">
+        <ComponentLabel text={_.startCase(currentController.kind)} />
```
I wonder if we shouldn't drop the labels in these rows. They seem fairly self-evident.

TheRealJon replied:
Follow on?

spadgett commented on frontend/public/components/project-overview.jsx:
```diff
@@ -0,0 +1,103 @@
+    }
+    <div className="project-overview__detail project-overview__detail--status">
+      <ComponentLabel text="Status" />
+      <Link to={`/k8s/ns/${namespace}/${kindObj.path}/${name}/pods`}>
```
Better to use `resourcePath` or `resourcePathForModel` so we keep the URL building in one place

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -68,33 +70,38 @@ class OverviewDetails extends React.Component {
-      const rightVersion = parseInt(_.get(right, 'openshift.io/deployment-config.latest-version'), 10);
+  sortReplicatorsByVersion(replicators, kind, descending) {
+    const compare = (left, right) => {
+      const annotation = kind === 'ReplicationController' ? 'openshift.io/deployment-config.latest-version' : 'deployment.kubernetes.io/revision';
```
I think it would be a little cleaner not to pass kind here, but to instead pass either the annotation name or a callback for getting the revision. Something like

```js
sortByRevision(replicators, descending, annotation) ...

sortRC(replicationControllers, descending) => sortByRevision(replicationControllers, descending, 'openshift.io/deployment-config.latest-version');

sortRS(replicaSets, descending) => sortByRevision(replicaSets, descending, 'deployment.kubernetes.io/revision');
```

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -135,24 +142,37 @@ class OverviewDetails extends React.Component {
-    const groups = _.groupBy(items, (item) => _.get(item, ['metadata', 'labels', label], 'other'));
+  groupItems(items, label) {
+    const compareGroups = (a, b) => {
+      if (a.name === 'other') {
```
This could potentially collide with a value since "other" is a legal label value

TheRealJon replied:
This issue goes beyond just this sorting function. In the case where a resource has a label value of 'other', it would be grouped with resources which don't have a value for the currently selected 'group by' label. Since labels can contain almost any string value, isn't the possibility of collision always going to be an issue?

TheRealJon replied:
Yeah, that would work to group them into a guaranteed unique group, but we would still have a problem with what to display as the heading for that group since we don't want that to collide either. We could use a value with a space in it which would make it an invalid label value, and it could still be used as a display value so that we don't have to do special logic when rendering. "No value" or something like that.

spadgett replied:
Let's look at it in a follow on

TheRealJon replied:
https://jira.coreos.com/browse/CONSOLE-755

spadgett commented on frontend/public/components/overview.jsx:
```diff
@@ -0,0 +1,372 @@
+  flags: PropTypes.object.isRequired
+};
+
+export const OverviewPage = connectToFlags(FLAGS.OPENSHIFT, FLAGS.PROJECTS_AVAILABLE)(OverviewPage_);
```
One drawback to this is you can't have private components. You have to export them all. It becomes more difficult to tell what's used internally and what's meant to be reused. It also discourages breaking things into small, simple components because it's a lot more work.

This would be a large refactoring effort and make it harder to backport fixes to previous releases as well because everything has moved. If you navigate to symbols instead of files in your editor, I feel like it shouldn't matter much what file the component is in.
